robot-server/robot_server/runs/run_store.py

```diff
@@ -62,20 +97,40 @@ def get(self, run_id: str) -> RunResource:
             run_id: Unique identifier of run entry to retrieve.
 
         Returns:
-            The retrieved run entry from the store.
+            The retrieved run entry from the db.
         """
-        try:
-            return self._runs_by_id[run_id]
-        except KeyError as e:
-            raise RunNotFoundError(run_id) from e
+        statement = sqlalchemy.select(run_table).where(run_table.c.id == run_id)
+        with self._sql_engine.begin() as transaction:
+            try:
+                row_run = transaction.execute(statement).one()
+            except sqlalchemy.exc.NoResultFound as e:
+                raise RunNotFoundError(run_id) from e
+
+            actions = _get_actions_no_transaction(run_id, transaction)
+            run = _convert_sql_row_to_run(
+                row_run,
+                [_convert_sql_row_to_action(action) for action in actions],
+                self._active_run,
+            )
+        return run
 
     def get_all(self) -> List[RunResource]:
         """Get all known run resources.
 
         Returns:
             All stored run entries.
         """
-        return list(self._runs_by_id.values())
+        statement = sqlalchemy.select(run_table)
+        with self._sql_engine.begin() as transaction:
+            runs = transaction.execute(statement).all()
+        return [
+            _convert_sql_row_to_run(
+                sql_row=row,
+                actions=self._get_actions(row.id),
+                current_run_id=self._active_run,
+            )
+            for row in runs
+        ]
```

Comment:
This looks like we're getting the run and its actions in two separate transactions. I think they need to be in a single transaction to provide a consistent snapshot view of everything.

(In practice, this probably doesn't actually matter right now because we're the only thing accessing this table, and all of these methods are synchronous. But I think it will matter more when they become `async`.)

We could restructure this like:

```suggestion
        statement = sqlalchemy.select(run_table)
        with self._sql_engine.begin() as transaction:
            runs = transaction.execute(statement).all()
            return [
                _convert_sql_row_to_run(
                    sql_row=row,
                    actions=self._get_no_transaction(row.id,  transaction),
                    current_run_id=self._active_run,
                )
                for row in runs
            ]
```

And then you could delete `self._get_actions()`, since this was the last thing using it.